test_contextual_lenses/test_learning.py

```diff
@@ -0,0 +1,454 @@
+"""End-to-end learning tests for contextual_lenses.py."""
+
+
+import jax.numpy as jnp
+
+import numpy as np
+
+from absl.testing import parameterized
+from absl.testing import absltest
+
+from contextual_lenses import mean_pool, max_pool, \
+linear_max_pool, linear_mean_pool, gated_conv 
+
+from train_utils import create_optimizer, train_step, \
+create_representation_model
+
+from encoders import one_hot_encoder, cnn_one_hot_encoder, \
+one_hot_pos_emb_encoder, cnn_one_hot_pos_emb_encoder
+
+
+def generate_random_sequences(bs=3, seq_len=12, num_categories=21):
```

Comment:
This function is also generating regression targets, but the function name and docstring don't reflect this. I'd pull the sampling of the random output_data into a different helper function.